sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -1036,12 +1036,8 @@ object SimplifyCasts extends Rule[LogicalPlan] {
 
   // Returns whether the from DataType can be safely casted to the to DataType without losing
   // any precision or range.
-  private def isWiderCast(from: DataType, to: NumericType): Boolean = (from, to) match {
-    case (from: NumericType, to: DecimalType) if to.isWiderThan(from) => true
-    case (from: DecimalType, to: NumericType) if from.isTighterThan(to) => true
-    case (from: IntegralType, to: IntegralType) => Cast.canUpCast(from, to)
-    case _ => from == to
```

Comment:
This refactor changes the behavior when `from` is non-numeric and `from == to`. Previously returns true, now false. Is this intentional?

To preserve behavior, and be more Scala-idiomatic by using pattern matching instead of `isInstanceOf`, I would suggest:
```scala
  private def isWiderCast(from: DataType, to: NumericType): Boolean = from match {
    case NumericType => Cast.canUpCast(from, to)
    case _ => from == to
}
```

Reply:
Previously, when `from` is non-numeric, `from == to` must be false.

Reply:
Do we have an existing test coverage for the case discussed in this thread?

Reply:
Previously, the existing test cases does't cover the case discussed above.
So, let's add one.